framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -1656,6 +1668,110 @@ public AnnotatedTypeMirror getAnnotatedTypeLhsNoTypeVarDefault(Tree lhsTree) {
+     */
+    private boolean computingAnnotatedTypeMirrorOfLHS = false;
+
+    /**
+     * Returns whether {@link #getAnnotatedTypeLhs(Tree)} is running right now. This controls which
+     * hierarchies' qualifiers are changed based on the receiver type and the declared annotations
+     * for a field.
+     *
+     * @return whether {@link #getAnnotatedTypeLhs(Tree)} is running right now
+     * @see #getAnnotatedTypeLhs(Tree)
+     */
+    public boolean isComputingAnnotatedTypeMirrorOfLHS() {
+        return computingAnnotatedTypeMirrorOfLHS;
+    }
+
+    /**
+     * Returns the type of a JavaExpression {@code expr} if it were evaluated before a tree {@code
+     * tree}.
+     *
+     * @param expr the expression to type.
+     * @param tree a tree.
+     * @return the type of {@code expr} if it were evaluated before tree {@code tree}.
+     */
+    public AnnotatedTypeMirror getAnnotatedTypeBefore(JavaExpression expr, ExpressionTree tree) {
+        CFAbstractStore<?, ?> store = getStoreBefore(tree);
+        CFAbstractValue<?> value = null;
+        if (CFAbstractStore.canInsertJavaExpression(expr)) {
+            value = store.getValue(expr);
+        }
+        Set<? extends AnnotationMirror> annos = null;
+        if (value != null) {
+            annos = value.getAnnotations();
+        } else {
+            // If there is no information in the store (possible if e.g., no refinement
+            // of the field has occurred), use top instead of automatically
+            // issuing a warning. This is not perfectly precise: for example,
+            // if jeExpr is a field it would be more precise to use the field's
+            // declared type rather than top. However, doing so would be unsound
+            // in at least three circumstances where the type of the field depends
+            // on the type of the receiver: (1) all fields in Nullness Checker,
+            // because of possibility that the receiver is under initialization,
+            // (2) polymorphic fields, and (3) fields whose type is a type variable.
+            // Using top here instead means that the method is always sound;
+            // a subclass can then override it with a more precise implementation.
+            annos = getQualifierHierarchy().getTopAnnotations();
+        }
+
+        AnnotatedTypeMirror res = AnnotatedTypeMirror.createType(expr.getType(), this, false);
+        res.addAnnotations(annos);
+        return res;
+    }
+
+    /**
+     * Returns the type of a left-hand side of an assignment.
+     *
+     * <p>The default implementation returns the type without considering dataflow type refinement.
+     * Subclass can override this method and add additional logic for computing the type of a LHS.
+     *
+     * @param lhsElement left-hand side of an assignment
+     * @return AnnotatedTypeMirror of {@code lhsElement}
+     */
+    public AnnotatedTypeMirror getAnnotatedTypeLhs(VariableElement lhsElement) {
```

Comment:
I don't understand what this method does. It take a `VariableElement` as parameter, that is, the declaration of a variable. So you don't know what particular assignment this is about, so you can't perform any type refinement. What am I missing?

Reply:
Ping on this.